runtime/src/builtins.rs

```diff
@@ -101,8 +94,63 @@ pub struct Builtins {
     /// Builtin programs that are always available
     pub genesis_builtins: Vec<Builtin>,
 
-    /// Builtin programs activated or deactivated dynamically by feature
-    pub feature_builtins: Vec<(Builtin, Pubkey, ActivationType)>,
+    /// Dynamic feature transitions for builtin programs
+    pub feature_transitions: Vec<BuiltinFeatureTransition>,
+}
+
+/// Actions taken by a bank when managing the list of active builtin programs.
+#[derive(Debug, Clone)]
+pub enum BuiltinAction {
+    Add(Builtin),
+    Remove(Pubkey),
+}
+
+/// State transition enum used for adding and removing builtin programs through
+/// feature activations.
+#[derive(Debug, Clone, AbiExample)]
+pub enum BuiltinFeatureTransition {
+    /// Add a builtin program if a feature is activated.
+    Add {
+        builtin: Builtin,
+        feature_id: Pubkey,
+    },
+    /// Remove a builtin program if a feature is activated or
+    /// retain a previously added builtin.
+    RemoveOrRetain {
+        previous_builtin: Builtin,
+        removal_feature_id: Pubkey,
+    },
+}
+
+impl BuiltinFeatureTransition {
+    pub fn to_action(
+        &self,
+        should_apply_action_for_feature: &impl Fn(&Pubkey) -> bool,
+    ) -> Option<BuiltinAction> {
+        match self {
+            Self::Add {
+                builtin,
+                feature_id,
+            } => {
+                if should_apply_action_for_feature(feature_id) {
+                    Some(BuiltinAction::Add(builtin.clone()))
+                } else {
+                    None
+                }
+            }
+            Self::RemoveOrRetain {
+                previous_builtin,
+                removal_feature_id,
+            } => {
+                if should_apply_action_for_feature(removal_feature_id) {
+                    Some(BuiltinAction::Remove(previous_builtin.id))
+                } else {
+                    // Retaining is no different from adding a new builtin.
+                    Some(BuiltinAction::Add(previous_builtin.clone()))
+                }
```

Comment:
I believe this branch introduced a new bug for epoch boundaries. On epoch boundaries, the `should_apply_action_for_feature(removal_feature_id)` condition will be false if the removal feature id wasn't newly activated. This causes the builtin to be re-added even if it was previously removed.